nondex-test/src/test/java/edu/illinois/nondex/core/MapTest.java

```diff
@@ -0,0 +1,127 @@
+/*
+The MIT License (MIT)
+Copyright (c) 2015 Alex Gyori
+Copyright (c) 2015 Owolabi Legunsen
+Copyright (c) 2015 Darko Marinov
+Copyright (c) 2015 August Shi
+
+
+Permission is hereby granted, free of charge, to any person obtaining
+a copy of this software and associated documentation files (the
+"Software"), to deal in the Software without restriction, including
+without limitation the rights to use, copy, modify, merge, publish,
+distribute, sublicense, and/or sell copies of the Software, and to
+permit persons to whom the Software is furnished to do so, subject to
+the following conditions:
+
+The above copyright notice and this permission notice shall be
+included in all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
+LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+*/
+
+package edu.illinois.nondex.core;
+
+import static org.hamcrest.core.IsEqual.equalTo;
+import static org.hamcrest.core.IsNot.not;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertThat;
+import static org.junit.Assert.assertTrue;
+
+import java.util.ConcurrentModificationException;
+import java.util.HashMap;
+import java.util.IdentityHashMap;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.NoSuchElementException;
+import java.util.WeakHashMap;
+import java.util.concurrent.ConcurrentHashMap;
+
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+import org.junit.runners.Parameterized.Parameters;
+
+@RunWith(Parameterized.class)
+public class MapTest<K, V> {
+    public Map<K, V> map;
+
+    public MapTest(Map<K, V> map) {
+        this.map = map;
+    }
+
+    @Parameters
+    public static Object[] data() {
+        HashMap<Integer, Integer> hm = new HashMap<>();
+        WeakHashMap<Integer, Integer> whm = new WeakHashMap<>();
+        IdentityHashMap<Integer, Integer> ihm = new IdentityHashMap<>();
+        ConcurrentHashMap<Integer, Integer> chm = new ConcurrentHashMap<>();
+        whm.clear();
+        ihm.clear();
+        chm.clear();
+        return new Object[] {hm};
+    }
+
+    @Before
+    public void setUp() {
+        for (int i = 0; i < 10; i++) {
+            ((Map<Integer, Integer>) map).put(i, i);
+        }
+    }
+
+    @Test(expected = IllegalStateException.class)
+    public void testRemoveWhenEmpty() {
+        map.clear();
```

Comment:
Add also a test to test that even though the map is not empty, if you remove without calling next this should still throw the exception.

Add also a test that if you call twice without next in between, it will throw exception, as per: " if the next method has not yet been called, or the remove method has already been called after the last call to the next method" as per: https://docs.oracle.com/javase/7/docs/api/java/util/Iterator.html#remove()